src/apps/js_v8/tmpl/kv_store.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#include "kv_store.h"
+
+#include "template.h"
+
+namespace ccf::v8_tmpl
+{
+  static TxContext* unwrap_tx_ctx(v8::Local<v8::Object> obj)
+  {
+    return static_cast<TxContext*>(obj->GetAlignedPointerFromInternalField(0));
+  }
+
+  static void js_kv_lookup(
+    v8::Local<v8::Name> name, const v8::PropertyCallbackInfo<v8::Value>& info)
+  {
+    if (name->IsSymbol())
+      return;
+
+    v8::Isolate* isolate = info.GetIsolate();
+    v8::Local<v8::Context> context = isolate->GetCurrentContext();
+    TxContext* tx_ctx_ptr = unwrap_tx_ctx(info.Holder());
+    std::string property_name = v8_util::to_str(isolate, name.As<v8::String>());
+
+    const auto [security_domain, access_category] =
+      kv::parse_map_name(property_name);
+
+    auto read_only = false;
+    switch (access_category)
+    {
+      case kv::AccessCategory::INTERNAL:
+      {
+        if (security_domain == kv::SecurityDomain::PUBLIC)
+        {
+          read_only = true;
+        }
+        else
+        {
+          v8_util::throw_error(
+            isolate,
+            fmt::format(
+              "JS application cannot access private internal CCF table '{}'",
+              property_name));
+          return;
+        }
+        break;
+      }
+      case kv::AccessCategory::GOVERNANCE:
+      {
+        read_only = tx_ctx_ptr->access != TxAccess::GOV_RW;
+        break;
+      }
+      case kv::AccessCategory::APPLICATION:
+      {
+        read_only = tx_ctx_ptr->access != TxAccess::APP;
+        break;
+      }
+      default:
+      {
+        v8_util::throw_error(
+          isolate,
+          fmt::format(
+            "Unhandled AccessCategory for table '{}'", property_name));
+        return;
+      }
+    }
+
+    auto handle = tx_ctx_ptr->tx->rw<KVMapType>(property_name);
+    v8::Local<v8::Value> value;
+    if (read_only)
+      value = KVMapReadOnly::wrap(context, *handle);
+    else
+      value = KVMapReadWrite::wrap(context, *handle);
+
+    info.GetReturnValue().Set(value);
+  }
+
+  v8::Local<v8::ObjectTemplate> KVStore::create_template(v8::Isolate* isolate)
+  {
+    v8::EscapableHandleScope handle_scope(isolate);
+
+    v8::Local<v8::ObjectTemplate> tmpl = v8::ObjectTemplate::New(isolate);
+
+    // Field 0: TxContext
+    tmpl->SetInternalFieldCount(1);
```

Comment:
Can we keep a static array of fields and take the size of it, or something a little more robust?

Reply:
Yes, good point. The proper solution would be to use an IDL and generated code like browsers do to avoid all this manual code, but unfortunately the one that Chromium has is highly specialized for their use cases. I'll put an enum in each file which should help.

Reply:
Done, I've added three new helper functions to set/get internal fields and set the field count using enum classes.

Reply:
I couldn't figure out how to make it nicer with structs so that you get a compile error when you forget to set a field. I think though it's not a big deal and ok as it is.

Reply:
Continued here: https://github.com/microsoft/CCF/pull/3258#discussion_r768080316
Marking as resolved.